Newly created groupId will delete automatically upon application shutdown.

cmabdullah opened the pull request:
A unique consumer group can be created for a Kafka listener by setting uniqueGroupId=true on the @KafkaListener annotation. This automatically generates a unique groupId each time the application starts. However, once the application shuts down, **_the generated groupId remains in Kafka and is not automatically deleted._**

```java
package org.cm.kafka;

import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;

@KafkaListener(groupId = "message-app", uniqueGroupId = true)
public class KafkaMessageConsumerClient {
	@Topic("user-online")
	void consumeEvent(String message) {
		System.out.println("event from Kafka: " + message);
	}
}

```
**This is the groupId list remains on kafka**

!['consumerGroupList'](https://i.imgur.com/dEO2NKV.png)
Multiple consumer groups support is added [on this](https://github.com/micronaut-projects/micronaut-kafka/pull/69).
where a flag is added (uniqueGroupId) in the annotation KafkaListener, if uniqueGroupId is true, _Then the app will create a unique groupID._

**However, the generated groupId persists in Kafka even after the application shuts down.**

We are currently developing a whiteboard application where each service have multiple instances, and any instance can broadcasts events, and all other instances need to receive these event. This follows a pub/sub pattern. 
As this feature is critical for our use case, we want to ensure better manageability by avoiding unnecessary groupId entries in Kafka. 

To address this, I have implemented a quick fix that still relies on annotations to delete newly created groupId will deleted automatically upon application shutdown.

I successfully extended the @KafkaListener annotation and tested the new feature using plain Kafka.

My initial idea is, I have added uniqueGroupIdDeleteOnShutdown field on @KafkaListener, **If uniqueGroupIdDeleteOnShutdown is set to true alongside uniqueGroupId, the unique group ID will be automatically deleted during the application shutdown process, ensuring better manageability of Kafka group IDs**.

```java
    @Override
    public void onApplicationEvent(ApplicationShutdownEvent event) {
        LOG.trace("Application shutdown initiated. Preparing to delete registered Kafka unique consumer groups.");
        List<String> uniqueGroupIdsDeleteOnShutdown = registerConsumerForGroupDeletion.values()
            .stream()
            .filter(consumerState -> consumerState.kafkaConsumer.groupMetadata() != null)
            .map(consumerState -> consumerState.kafkaConsumer.groupMetadata().groupId())
            .toList();
        if (!uniqueGroupIdsDeleteOnShutdown.isEmpty()) {
            LOG.trace("Closing {} consumers and attempting to delete the following consumer groups: {}",
                uniqueGroupIdsDeleteOnShutdown.size(), uniqueGroupIdsDeleteOnShutdown);
            closeConsumers();
            adminClient.deleteConsumerGroups(uniqueGroupIdsDeleteOnShutdown)
                .all().whenComplete((voidResult, throwable) -> {
                    if (throwable == null) {
                        LOG.trace("Successfully deleted the following consumer groups: {}", uniqueGroupIdsDeleteOnShutdown);
                    } else {
                        LOG.error("Failed to delete the following consumer groups: {}. Error: {}",
                            uniqueGroupIdsDeleteOnShutdown, throwable.getMessage(), throwable);
                    }
                });
        } else {
            LOG.trace("No unique consumer groups are registered for deletion.");
        }
    }
```

cmabdullah committed:
Delete newly created groupId will delete automatically upon application shutdown.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/micronaut-projects/micronaut-kafka?pullRequest=1129) <br/>All committers have signed the CLA.

renovate[bot] committed:
chore(deps): update dependency gradle to v8.12 (#1111)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency io.micronaut.micrometer:micronaut-micrometer-bom to v5.9.3 (#1109)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

graemerocher commented:
can you add a test please?

iNviNho commented:
This is a nice addition (I haven't checked the code yet) but please bear in mind that not all applications will be granted access to delete consumer group.

If this is meant to go through, I would either make this deletion by default disabled or make the log level only warning or info as this is not critical functionality that has to work.

Byt the way you can always decrease retention on internal topic that you use to store committed offset and therefore such unused consumer groups will eventually disappear.

graemerocher commented:
it should absolutely be opt-in

cmabdullah committed:
Newly created groupId will delete automatically upon application shutdown.
- update log level

cmabdullah committed:
Newly created groupId will delete automatically upon application shutdown.
- Managed List of unique consumer group IDs, as accessing consumerState.kafkaConsumer.groupMetadata() is not the best way; the consumer is running by another thread through the executor service, so it should not be a good way to access consumerState.kafkaConsumer.groupMetadata() from another thread.

cmabdullah committed:
Newly created groupId will delete automatically upon application shutdown.
- added test case for the feature groupId will delete automatically upon application shutdown

cmabdullah committed:
Newly created groupId will delete automatically upon application shutdown.
- revert to old state

cmabdullah commented:
@graemerocher, 
Thank you for the feedback! 
I have added an integration test as requested and also provided a video demonstration at the given [link](https://drive.google.com/drive/folders/1t2KPCN6_C0LejppUKvXV7H4NNmdRtG_G?usp=sharing). 

Additionally, I’ve made the deletion functionality disabled by default, ensuring it aligns with the concerns raised.

I’m also sharing the [build log ](https://github.com/cmabdullah/KafkaConsumerPOC/blob/main/README.md)for reference. Please let me know if any further refinements are needed. Looking forward to your thoughts!

cmabdullah commented:
Hi @sdelamo Whenever you get a chance, I’d appreciate your feedback. Let me know if any improvements are needed.

Thanks!

sdelamo commented on kafka/src/main/java/io/micronaut/configuration/kafka/annotation/KafkaListener.java:
```diff
@@ -65,6 +65,33 @@
+     * }</pre>
+     *
+     * @return {@code true} to delete the unique group ID on shutdown, {@code false} otherwise.
+     * Defaults to {@code false}.
```
```suggestion
     * Defaults to {@code false}.
     * @since 5.8.0
```

sdelamo committed:
Update kafka/src/main/java/io/micronaut/configuration/kafka/annotation/KafkaListener.java
